crates/curp/src/client/stream.rs

```diff
@@ -0,0 +1,113 @@
+        loop {
+            let Some(leader_id) = self.state.leader_id().await else {
+                debug!("cannot find the leader id in state, wait for leadership update");
+                self.state.leader_notifier().listen().await;
+                continue;
+            };
+            if let Some(local_id) = self.state.local_server_id() {
+                if leader_id == local_id {
+                    self.state.check_gen_local_client_id().await;
+                    debug!("skip keep heartbeat for local connection, wait for leadership update");
+                    self.state.leader_notifier().listen().await;
+                    continue;
+                }
+            }
+            return self.state.map_server(leader_id, f).await;
+        }
+    }
+
+    /// Keep heartbeat
+    pub(super) async fn keep_heartbeat(&self) {
+        /// Prevent lock contention when leader crashed or some unknown errors
+        const RETRY_DELAY: Duration = Duration::from_millis(100);
+
+        loop {
+            // is heartbeat task cancellation safety?
+            let heartbeat = self.map_remote_leader::<(), _>(|conn| async move {
+                loop {
+                    let err = conn
+                        .lease_keep_alive(
+                            self.state.clone_client_id(),
+                            self.config.heartbeat_interval,
+                        )
+                        .await;
+                    #[allow(clippy::wildcard_enum_match_arm)]
+                    match err {
+                        CurpError::Redirect(Redirect { leader_id, term }) => {
+                            let _ig = self.state.check_and_update_leader(leader_id, term).await;
+                        }
+                        CurpError::WrongClusterVersion(_) => {
+                            warn!(
+                                "cannot find the leader in connects, wait for  leadership update"
+                            );
+                            self.state.leader_notifier().listen().await;
+                        }
+                        CurpError::ShuttingDown(_) => {
+                            debug!("shutting down stream client background task");
+                            break Err(err);
+                        }
+                        _ => {
+                            warn!("got unexpected error {err:?} when keep heartbeat, retrying...");
+                            tokio::time::sleep(RETRY_DELAY).await;
+                        }
+                    }
+                }
+            });
+
+            tokio::select! {
+                _ = self.state.leader_notifier().listen() => {
+                    debug!("interrupt keep heartbeat because leadership changed");
+                    // TODO release the heartbeat task unless cancellation safety is ensured, especially in the case of the tonic RPC method.
+                    // TODO how to release?
+                }
+                _ = heartbeat => {
```

Comment:
The heartbeat error seems get ignored?

Reply:
Yes, It seems that the only possible value (`ShuttingDown` error) does not need to be handled at the moment.